### ISSUE TITLE
Non-deterministic type error

### ISSUE THREAD

**Issue by smackesey** [microsoft/pyright#6679]
This error occasionally shows up when running `pyright` on [Dagster](https://github.com/dagster-io/dagster):

```
[2023-12-07T19:58:15Z]   143:29: Argument of type "Unknown \| Mapping[str, Any] \| None" cannot be assigned to parameter "container_config" of type "Mapping[str, Any]" in function "__new__"
  | [2023-12-07T19:58:15Z]   Type "Unknown \| Mapping[str, Any] \| None" cannot be assigned to type "Mapping[str, Any]"
  | [2023-12-07T19:58:15Z]     "None" is incompatible with "Mapping[str, Any]" (reportGeneralTypeIssues)
```

But then goes away on a re-run. This has happened with both Pyright 1.1.332 and 1.1.339.

Here is the flagged line in the code: https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-k8s/dagster_k8s/job.py#L143:

```
        return super(UserDefinedDagsterK8sConfig, cls).__new__(
            cls,
->          container_config=container_config,
            pod_template_spec_metadata=pod_template_spec_metadata,
            pod_spec_config=pod_spec_config,
            job_config=job_config,
            job_metadata=job_metadata,
            job_spec_config=job_spec_config,
            merge_behavior=check.inst_param(
                merge_behavior, "merge_behavior", K8sConfigMergeBehavior
            ),
        )
```

I'm not totally sure what's going on but I think the `Unknown` for `container_config` is coming from the function `k8s_snake_case_dict` that lacks a return type annotation. Earlier in the code there is this:

```
        if container_config:
            container_config = k8s_snake_case_dict(kubernetes.client.V1Container, container_config)

```

Maybe there is a race condition between type analysis on the code above and inference for that function.

**VS Code extension or command-line**

Command line (1.1.332 <-> 1.1.339)

**Comment by erictraut** [comment]
When you say that "it occasionally shows up", do you mean that you can run the same version of pyright with the sam configuration on the same system multiple times and it gives you different results? Or is there something else that changes — the version of pyright, the code in the source base, etc.?

Pyright is single-threaded, so there aren't many opportunities for race conditions. Non-deterministic behavior can sometimes occur when you're using interactive language server features because they affect the order in which types are evaluated, but it sounds like you're seeing this with the CLI.

**Comment by smackesey** [comment]
>When you say that "it occasionally shows up", do you mean that you can run the same version of pyright with the sam configuration on the same system multiple times and it gives you different results? Or is there something else that changes — the version of pyright, the code in the source base, etc.?

Yes-- exact same code, same version of pyright.

**Comment by erictraut** [comment]
I have a partial theory, but I haven't been able to put all the pieces together yet. I suspect that there's a behavior that depends on the order in which modules are type checked. That would mean that the results would depend on the order in which files are enumerated by the file system, which can be somewhat unpredictable depending on the file system and the order in which the files were created and updated.

**Comment by smackesey** [comment]
Just a heads up that if you try to replicate the bug in the current master of `dagster`, I suspect it won't surface. I just committed some missing annotations (specifically on the `k8s_snake_case_dict` function I mention above) that I suspect will fix the issue. If you want to try replicating the last commit before these annotations were added is https://github.com/dagster-io/dagster/commit/148f49a7f574e468f806e1e932bc61ed5feeea24

**Comment by erictraut** [comment]
I still haven't figured out why you're seeing non-deterministic behavior. When I run the code, it always produces the error you mentioned in your bug report. That error is expected if the call to `k8s_snake_case_dict` evaluates to `Unknown`. When an `Unknown` value is assigned to a variable with a declared type, allows this assignment and "narrows" the resulting type of the variable to `Unknown | <declared type>`. In your case, the declared type of `container_config` is `Mapping[str, Any] | None`, so the resulting type is `Unknown | Mapping[str, Any] | None`. This type isn't compatible with the `container_config` parameter in the `super().__new__` call, so a type violation should be reported here.

I was surprised to see that the call to `k8s_snake_case_dict` evaluated to `Unknown`. It turns out there was a bug in the return type inference logic. If `analyzeUnannotatedFunctions` was configured to "false", it never inferred return types regardless of whether the function contained (some) input parameter annotations. In the case of `k8s_snake_case_dict`, one of the two input parameters is annotated, so pyright analyzes the the code in this function. It's reasonable, therefore, for pyright to infer the return type.

I suspect that you're one of very few pyright users who set `analyzeUnannotatedFunctions` to false. I've never seen anyone use this before in production. The only reason I added the option was to enable me to do apples-to-apples performance comparisons between pyright and mypy. As you know, mypy skips analysis of unannotated functions by default.

With the bug fix I mentioned above, this will no longer occur for you. It's possible that there's still a bug lurking here that leads to non-determinism, but I've run out of theories, and without a repro case, I'm not going to spend any further time on it. If you see any non-deterministic behaviors in the future, please report them because that might give me additional clues.

**Comment by smackesey** [comment]
Great, thanks for fixing the `analyzeUnannotatedFunctions` bug. We've been using pyright for about a year and this is the first time I've seen any non-determinism so I'm not holding my breath.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.340, which I just published.
